Updates to display names

fivetran-reneeli opened the pull request:
## PR Overview
**This PR will address the following Issue/Feature:**
Internal ticket

**This PR will result in the following new package version:**
<!--- Please add details around your decision for breaking vs non-breaking version upgrade. If this is a breaking change, were backwards-compatible options explored? -->
v0.3.0

Breaking change due to removal of some fields

**Please provide the finalized CHANGELOG entry which details the relevant changes included in this PR:**
<!--- Copy/paste the CHANGELOG for this version below. -->


## 🚨 Breaking Changes
-  The following fields were removed from the `servicenow__task_enhanced` model as `problem_first_reported_by_task_value` was incorrectly assumed to have corresponding `sys_user` values, but in actuality is related to the task table and therefore have no user fields.
  - `problem_reporter_email, problem_reporter_manager_value, problem_reporter_department_value, problem_reporter_name, problem_reporter_roles`

## Feature Updates
- The following fields were added to the `servicenow__incident_enhanced` model to add relevant display names to the caller_id_value.
  - `caller_email, caller_manager_value, caller_department_value, caller_name, caller_roles`
- The following fields were added to the `servicenow__user_enhanced` model to add relevant display names to the company_value and manager_value. 
  - `company_name, manager_name`
- The following label fields were added to respective choice fields in order to bring the human-readable text displayed for each choice option:
  - `priority_label, impact_label, urgency_label, task_state_label, problem_state_label, change_request_type_label, change_request_phase_state_label, change_request_risk_label, risk_impact_analysis_label, incident_category_label, incident_subcategory_label, incident_severity_label, incident_state_label, business_impact_label`


## PR Checklist
### Basic Validation
Please acknowledge that you have successfully performed the following commands locally:
- [x] dbt run –full-refresh && dbt test
- [x] dbt run (if incremental models are present) && dbt test

Before marking this PR as "ready for review" the following have been applied:
- [na] The appropriate issue has been linked, tagged, and properly assigned.
- [x] All necessary documentation and version upgrades have been applied.
    <!--- Be sure to update the package version in the dbt_project.yml, integration_tests/dbt_project.yml, and README if necessary. -->
- [x] docs were regenerated (unless this PR does not include any code or yml updates).
- [x] BuildKite integration tests are passing.
- [x] Detailed validation steps have been provided below.

### Detailed Validation
Please share any and all of your validation steps:
<!--- Provide the steps you took to validate your changes below. -->

Other than checking for fanouts using the unique tests in the servicenow.yml, additional validation tests were added and ran as below: 

![image](https://github.com/user-attachments/assets/25706b9c-cfc5-45b8-a261-56efa0caeb2d)

The `user_aggregated_aggs` test is failing since we added new fields and so are those are not available in prod yet.

No variables were set other than    ` fivetran_validation_tests_enabled: true`

### If you had to summarize this PR in an emoji, which would it be?
<!--- For a complete list of markdown compatible emojis check our this git repo (https://gist.github.com/rxaviers/7360908)  --> 
:dancer:

fivetran-joemarkiewicz commented on CHANGELOG.md:
```diff
@@ -1,3 +1,20 @@
+
+## 🚨 Breaking Changes
+-  The following fields were removed from the `servicenow__task_enhanced` model as `problem_first_reported_by_task_value` was incorrectly assumed to have corresponding `sys_user` values, but in actuality is related to the task table and therefore have no user fields.
+  - `problem_reporter_email, problem_reporter_manager_value, problem_reporter_department_value, problem_reporter_name, problem_reporter_roles`
```
Can you make these their own bullets to make it clear what fields were removed instead of adding them all in a single code block. For example, I'm thinking of the following format:
```suggestion
  - `problem_reporter_email`
  - `problem_reporter_manager_value`
  - `problem_reporter_department_value`
  - `problem_reporter_name`
  - `problem_reporter_roles`
```

fivetran-joemarkiewicz replied:
Can you make the same update to the similar entries in this CHANGELOG that list out field additions/removals.

fivetran-reneeli replied:
Updated the format!

fivetran-joemarkiewicz commented on CHANGELOG.md:
```diff
@@ -1,3 +1,20 @@
+-  The following fields were removed from the `servicenow__task_enhanced` model as `problem_first_reported_by_task_value` was incorrectly assumed to have corresponding `sys_user` values, but in actuality is related to the task table and therefore have no user fields.
+  - `problem_reporter_email, problem_reporter_manager_value, problem_reporter_department_value, problem_reporter_name, problem_reporter_roles`
+
+## Feature Updates
```
A new table was added to the staging models. We should be calling that out in the feature updates as well.

fivetran-reneeli replied:
added

fivetran-joemarkiewicz commented on models/docs.md:
```diff
@@ -1931,3 +1931,51 @@ List of included roles.
 {% enddocs %}
+
+{% docs sys_choice %}
+Staging table containing all records relating to system choices. System choices represent predefined options for fields in various ServiceNow tables.
```
This is the definition of the source table. I recommend removing the staging wording from this description.

```suggestion
Table containing all records relating to system choices. System choices represent predefined options for fields in various ServiceNow tables.
```

fivetran-joemarkiewicz commented on models/servicenow.yml:
```diff
@@ -379,6 +382,18 @@ models:
       - name: is_change_request_unauthorized
         description: '{{ doc("is_change_request_unauthorized") }}'
 
+      - name: dv_priority_label
+        description: '{{ doc("label") }}'
+
+      - name: dv_impact_label
+        description: '{{ doc("label") }}'
+
+      - name: dv_urgency_label
+        description: '{{ doc("label") }}'
+
+      - name: dv_task_state_label
+        description: '{{ doc("label") }}'
```
It could be confusing that all of these will have the same definition. Is there a way we can work the unique name into each of these field descriptions?

fivetran-joemarkiewicz replied:
This comment goes for all the other label descriptions as well.

fivetran-joemarkiewicz replied:
For example, I don't necessarily know what `dv_` stands for here.

fivetran-reneeli replied:
I don't know about listing separate definitions for each field since it'll be mostly redundant. Instead, I updated the definition for label to be "The display value, or human-readable text, for a **titular** choice option." That way the reader knows it's the label for the field of topic.

Also regarding `dv_`, that prefix was added by direct mention in the customer's request. I went back and forth on it since it's not very intuitive what it stands for. I did update the field description to mention it so hopefully it's clearer. I'm not opposed to just removing it entirely though.

fivetran-joemarkiewicz commented on models/servicenow.yml:
```diff
@@ -1031,6 +1049,21 @@ models:
       - name: caller_id_value
         description: '{{ doc("caller_id_value")}}'
+
+      - name: caller_email
+        description: '{{ doc("email")}}'
+
+      - name: caller_manager_value
+        description: '{{ doc("manager_value")}}'
+
+      - name: caller_department_value
+        description: '{{ doc("department_value")}}'
+
+      - name: caller_name
+        description: '{{ doc("sys_user_name")}}'
+
+      - name: caller_roles
+        description: '{{ doc("roles")}}'
```
Should we make it clear in the definitions that these are for the caller? I know the name of the field mentions caller, but it may be confusing with the definition of the field matching another field with a different name.

fivetran-joemarkiewicz replied:
Same comment for other types of fields that have this definition approach.

fivetran-reneeli replied:
I feel like it'll get repetitive, especially since the person that the field is regarding is stipulated in the field name itself. But I understand we can make it clearer. Perhaps instead of, for example, `department_value` = `Value representing the department associated with the system user.`, I can use the fact that the field name includes the user's role and update the description to 
`Value representing the department associated with the titular user.` That way the reader knows it's regarding the user in the field.

Let me know what you think, and will make the updates if that sounds good!

fivetran-joemarkiewicz commented on models/staging/stg_servicenow__change_request.sql:
```diff
@@ -53,8 +52,8 @@ final as (
         outside_maintenance_schedule,
-        phase as change_request_phase,
-        phase_state as change_request_phase_state,
+        cast(phase as {{ dbt.type_string() }}) as change_request_phase,
```
These small cast changes should also be documented in the CHANGELOG. We don't need to be 100% specific to what fields were changed, but we should still make a clear callout that these changes were made.

fivetran-joemarkiewicz replied:
Same comment for similar changes in staging models.

fivetran-reneeli replied:
updated!

fivetran-reneeli commented on DECISIONLOG.md:
```diff
@@ -1,3 +1,11 @@
+## Methodology for Adding Labels for Choice Fields 
```
updaetd the CHANGELOG

fivetran-reneeli commented:
> @fivetran-reneeli thanks for opening this PR. I was able to review and have a few requests.
> 
> Additionally, would you be able to provide validation tests for the changed end models and share any other validations. I imagine integrity tests would be especially valuable here since you were initially encountering fanout issues when iterating through these changes.

Hi @fivetran-joemarkiewicz , this is ready for re-review. Regarding testing, I also initially relied on just the unique tests in yml to check for fanouts, but now have added the validation tests and updated the template accordingly.

fivetran-joemarkiewicz reviewed (changes requested):
@fivetran-reneeli thanks for working through these updates! The code generally looks good, I just have a few final notes before this will be ready for approval. Let me know if you have any questions.

fivetran-reneeli commented on models/servicenow__problem_enhanced.sql:
```diff
@@ -122,6 +122,11 @@ problem_enhanced as (
         and problem.source_relation = problem_workaround_communicator.source_relation
+    left join sys_choice state_choice
+        on problem.problem_state = state_choice.sys_choice_value
+        and state_choice.sys_choice_name = 'problem'
```
Ah I forgot to move the filter to the initial CTE. Updated here and the rest of the models too.

fivetran-reneeli commented on integration_tests/tests/integrity/change_request_match.sql:
```diff
@@ -0,0 +1,33 @@
+    enabled=var('fivetran_validation_tests_enabled', false)
+) }}
+
+-- The necessary source and source_filter adjustments used below originate from the int_zendesk__sla_policy_applied model
```
Oops forgot to remove those comments~ done

fivetran-reneeli commented on models/docs.md:
```diff
@@ -1931,3 +1931,51 @@ List of included roles.
+{% enddocs %}
+
+{% docs label %}
+The display value, or human-readable text, for a titular choice option.
```
updated

fivetran-reneeli commented on models/servicenow__task_enhanced.sql:
```diff
@@ -312,7 +301,27 @@ left join cmdb_ci
   and task.source_relation = business_service.source_relation
+left join sys_choice state_choice
+  on task.task_state = state_choice.sys_choice_value
+  and state_choice.sys_choice_name = 'task'
```
moved to initial CTE

fivetran-reneeli commented on integration_tests/tests/integrity/change_request_match.sql:
```diff
@@ -0,0 +1,33 @@
+-- The necessary source and source_filter adjustments used below originate from the int_zendesk__sla_policy_applied model
+with source as (
+    select
+        1 as join_key,
```
Ah, I was thinking to just do overall row count match, but that's true if we want to be extra sure that there are no fanouts at each grain. I will try that instead.

fivetran-reneeli commented on models/servicenow__incident_enhanced.sql:
```diff
@@ -66,6 +80,31 @@ incident_enhanced as (
+    left join sys_choice business_impact_choice
+        on incident.business_impact = business_impact_choice.sys_choice_value
+        and business_impact_choice.element = 'business_impact'
+        -- none when filtered for incident
```
Ah yeah so I found that this label doesn't exist in our sys_choice table for this set of filters. Being as this is the only occurrence, I just assume this may be just due to a limitation on our end. I should remove that comment

![image](https://github.com/user-attachments/assets/803f0e66-cd02-40e3-9b38-e54b96d5ab7b)

fivetran-reneeli commented on CHANGELOG.md:
```diff
@@ -1,3 +1,51 @@
+- For more information about the logic used to attach labels to choice fields, refer to the [DECISIONLOG](https://github.com/fivetran/dbt_servicenow/blob/main/DECISIONLOG.md#methodology-for-adding-label-for-choice-fields) where a section has been added (titled *Methodology for Adding Labels for Choice Fields*).
+
+#### New Staging Model
+- To make labels feasible, we included a new table to the staging models, `stg_servicenow__sys_choice`.
```
good point!

fivetran-reneeli commented on integration_tests/dbt_project.yml:
```diff
@@ -72,4 +73,9 @@ seeds:
+
+
+  ## For validation testing. To be commented out before release.
+  # servicenow_schema: servicenow_test_env  
```
oops I think this was leftover from something, just deleted

fivetran-reneeli commented on integration_tests/tests/consistency/activity_enhanced_aggs.sql:
```diff
@@ -0,0 +1,93 @@
+with prod as (
+    select
+        task_updated_date,
+        -- task_state,
+        -- priority,
+        -- impact,
+        -- urgency,
```
Hm yeah I wasn't sure what to place the grain as, the date or by each of these dimensions. So I left it there as an option. I suppose for this purpose, comparing by date should be fine so I'll remove it

fivetran-reneeli commented on models/servicenow__user_enhanced.sql:
Yes can do-- not sure how else to go about validating it though other than doing a row count since this model doesn't have counts like the `activity_summary` does. I'm thinking I can do a test on counts of the following fields


        count_distinct_sys_user_group_ids,
        count_distinct_sys_user_role_ids,
        count_distinct_sys_user_role_names,
        count_distinct_included_roles,
        count_distinct_sys_user_group_roles

which isn't actually in the final CTE as it's in an intermediate CTE in the upstream [user_aggregates](https://github.com/fivetran/dbt_servicenow/blob/2893a89918444896d0ae485c87f5782b26dddb9c/models/servicenow__user_aggregated.sql#L38-L42) model. But there may be value in popping these out to include as well in the final user model. 

Let me know though if you think a row count test is fine and if above is overkill.

I suppose I can also do a match on the [string aggs](https://github.com/fivetran/dbt_servicenow/blob/2893a89918444896d0ae485c87f5782b26dddb9c/models/servicenow__user_aggregated.sql#L66-L85) too.

fivetran-reneeli commented on integration_tests/dbt_project.yml:
```diff
@@ -26,6 +26,7 @@ vars:
     servicenow_sys_user_has_role_identifier: "sys_user_has_role_data"
     servicenow_sys_user_grmember_identifier: "sys_user_grmember_data"
+    # fivetran_validation_tests_enabled: true
 
```
done!

fivetran-reneeli commented on integration_tests/dbt_project.yml:
```diff
@@ -26,6 +26,7 @@ vars:
     servicenow_sys_user_has_role_identifier: "sys_user_has_role_data"
     servicenow_sys_user_grmember_identifier: "sys_user_grmember_data"
+    # fivetran_validation_tests_enabled: true
 
```
Ah right

fivetran-reneeli commented on integration_tests/dbt_project.yml:
```diff
@@ -26,6 +26,7 @@ vars:
     servicenow_sys_user_role_identifier: "sys_user_role_data"
     servicenow_sys_user_has_role_identifier: "sys_user_has_role_data"
     servicenow_sys_user_grmember_identifier: "sys_user_grmember_data"
+    # fivetran_validation_tests_enabled: true
```
removed

fivetran-reneeli commented on integration_tests/tests/consistency/activity_enhanced_aggs.sql:
```diff
@@ -0,0 +1,93 @@
+select * 
+from final
+    where prod.task_updated_date != dev.task_updated_date or 
+    prod_source_relation != dev_source_relation or
```
Good callout thank you! I made some significant updates to the validation tests

fivetran-reneeli commented on integration_tests/tests/consistency/activity_enhanced_aggs.sql:
Added

fivetran-reneeli commented:
Thanks Joe for the notes! I've addressed and also made a lot of adjustments in the validation tests and updated the template accordingly with the run results. Let me know what you think

fivetran-joemarkiewicz commented on integration_tests/dbt_project.yml:
```diff
@@ -27,6 +27,10 @@ vars:
 
+    servicenow__using_roles: true # to be turned on when generating docs so that can be generated via the manifest and docs as it is turned off by default.
+
+# models:
+#   +schema: "servicenow_{{ var('directed_schema','dev') }}"
```
We can keep this uncommented. It won't impact the integration test runs.

fivetran-reneeli replied:
sure thing!

fivetran-joemarkiewicz commented on integration_tests/tests/consistency/change_request_enhanced_aggs.sql:
Small request, can we change the names of these consitency tests which are testing the row count to explicitly be named `*_row_count`. For example, the name of this test implies we are validating various aggregates (like we are in the activity_enhaned_aggs validation. However, for these we are only testing the row count. Can you make these name updates to the relevant consistency tests.

fivetran-reneeli replied:
Updated1

fivetran-joemarkiewicz commented on integration_tests/tests/consistency/user_aggregated_aggs.sql:
```diff
@@ -0,0 +1,63 @@
+        count_distinct_included_roles,
+        count_distinct_sys_user_group_roles
+    from {{ target.schema }}_servicenow_prod.servicenow__user_aggregated
+    group by 1,2
```
Even after prod is in sync with dev this test will still fail. We are grouping by `sys_user_id` and `source_relation`, but then not doing any aggregates. Therefore we result in an error like the following once all fields are available in prod and dev.

![image](https://github.com/user-attachments/assets/6077f256-32fc-4425-8a4c-23655ce67722)

We need to adjust this test to ensure it will work once the changes are live. I imagine doing a sum on these aggregate fields should do the trick.

fivetran-reneeli replied:
Ah we don't need the `group by` here as it's already aggregated on the `sys_user_id` level from the model. Removed that

fivetran-joemarkiewicz commented on integration_tests/tests/consistency/user_aggregated_aggs.sql:
```diff
@@ -0,0 +1,63 @@
+        dev.sys_user_id,
+        prod.source_relation,
+        dev.source_relation,
+        prod.count_distinct_sys_user_group_ids,
+        dev.count_distinct_sys_user_group_ids,
+        prod.count_distinct_sys_user_role_ids,
+        dev.count_distinct_sys_user_role_ids,
+        prod.count_distinct_sys_user_role_names,
+        dev.count_distinct_sys_user_role_names,
+        prod.count_distinct_included_roles,
+        dev.count_distinct_included_roles,
+        prod.count_distinct_sys_user_group_roles,
+        dev.count_distinct_sys_user_group_roles
```
You'll also want to differentiate the fields for dev and prod. Otherwise the `abs` where clause you have in the final select query will not know what to reference.

fivetran-reneeli replied:
Wow what an oversight, thanks for noticing! 🤯

fivetran-joemarkiewicz commented on models/servicenow__user_aggregated.sql:
```diff
@@ -35,11 +35,11 @@ aggregate_counts as (
-        count(distinct user_role.sys_user_role_name) as count_distinct_sys_user_role_names,
-        count(distinct user_role.includes_roles) as count_distinct_included_roles,
-        count(distinct user_group.sys_user_group_roles) as count_distinct_sys_user_group_roles
+        count(distinct case when user_grmember.sys_user_group_id is not null and user_grmember.sys_user_group_id != '' then user_grmember.sys_user_group_id else null end) as count_distinct_sys_user_group_ids,
+        count(distinct case when user_has_role.sys_user_role_id is not null and user_has_role.sys_user_role_id != '' then user_has_role.sys_user_role_id else null end) as count_distinct_sys_user_role_ids,
+        count(distinct case when user_role.sys_user_role_name is not null and user_role.sys_user_role_name != '' then user_role.sys_user_role_name else null end) as count_distinct_sys_user_role_names,
+        count(distinct case when user_role.includes_roles is not null and user_role.includes_roles != '' then user_role.includes_roles else null end) as count_distinct_included_roles,
+        count(distinct case when user_group.sys_user_group_roles is not null and user_group.sys_user_group_roles != '' then user_group.sys_user_group_roles else null end) as count_distinct_sys_user_group_roles
```
These case when statements look like recent updates. Would you be able to explain these changes? Why are we also checking a few fields in other tables to see if they are empty strings? 

Also, should we document this change in the CHANGELOG since it could impact the results from previous versions?

fivetran-joemarkiewicz replied:
Also be aware that these new fields are also brought into the `user_enhanced` model because of the select * on this dependent model. So we will need to update the CHANGELOG documentation to reflect this change as well.

fivetran-reneeli replied:
Yep, so this change was because I realized the original counts were incorrect due to unintentionally counting blank values, that were not null. I've added an example at the bottom of this [hex notebook](on aggregate_counts.sys_user_id = user_aggregates.sys_user_id). 

Therefore this `case when` update ensures those blank values are not included in the counts!

I will add this to the CHANGELOG! I've made sure they're in the ymls.

fivetran-joemarkiewicz reviewed (changes requested):
@fivetran-reneeli this is really close to approval. Just a few final comments to be addressed. Let me know if you have any questions.

fivetran-joemarkiewicz reviewed (approved):
@fivetran-reneeli thanks for working through this PR and my comments. I have one super small CHANGELOG request to remove the emoji, then this will be ready for release review. Thanks!

fivetran-reneeli commented on CHANGELOG.md:
```diff
@@ -1,3 +1,60 @@
+[PR #10](https://github.com/fivetran/dbt_servicenow/pull/10) includes the following updates:
+
+## Breaking Changes
+-  The following fields were removed from the `servicenow__task_enhanced` model as `problem_first_reported_by_task_value` was incorrectly assumed to have corresponding `sys_user` values, but in actuality is related to the task table and therefore have no user fields.
```
Yes you're spot on!

fivetran-reneeli commented on CHANGELOG.md:
```diff
@@ -1,3 +1,60 @@
+## Under The Hood
+- Added explicit casts to timestamp fields, as well as string casts to choice fields in order to join them later downstream on `sys_choice.element`.
+- Added consistency and integrity validation tests within integration_tests folder for all current end models.
+- Updated the logic for 
```
Oops-- I had made a new section instead and forgot to delete that-- removed

fivetran-catfritz reviewed (approved):
lgtm for release!

fivetran-catfritz reviewed (changes requested):
Sorry I just noticed the versions in the README need to be updated.
